src/Swarm/Language/Syntax.hs

```diff
@@ -85,51 +91,42 @@ import Swarm.Language.Types
 
 -- | The type of directions. Used /e.g./ to indicate which way a robot
 --   will turn.
-data Direction = Lft | Rgt | Back | Fwd | North | South | East | West | Down
-  deriving (Eq, Ord, Show, Read, Generic, Data, Hashable, ToJSON, FromJSON)
+data Direction = DLeft | DRight | DBack | DForward | DNorth | DSouth | DEast | DWest | DDown
+  deriving (Eq, Ord, Show, Read, Generic, Data, Hashable, ToJSON, FromJSON, Enum, Bounded)
 
 instance ToJSONKey Direction where
   toJSONKey = genericToJSONKey defaultJSONKeyOptions
 
 instance FromJSONKey Direction where
   fromJSONKey = genericFromJSONKey defaultJSONKeyOptions
 
--- | The 'applyTurn' function gives the meaning of each 'Direction' by
---   turning relative to the given vector or by turning to an absolute
---   direction vector.
-applyTurn :: Direction -> V2 Int64 -> V2 Int64
-applyTurn Lft (V2 x y) = V2 (- y) x
-applyTurn Rgt (V2 x y) = V2 y (- x)
-applyTurn Back (V2 x y) = V2 (- x) (- y)
-applyTurn Fwd v = v
-applyTurn North _ = north
-applyTurn South _ = south
-applyTurn East _ = east
-applyTurn West _ = west
-applyTurn Down _ = V2 0 0
-
--- | Possibly convert a vector into a 'Direction'---that is, if the
---   vector happens to be a unit vector in one of the cardinal
---   directions.
-toDirection :: V2 Int64 -> Maybe Direction
-toDirection v = case v of
-  V2 0 1 -> Just North
-  V2 0 (-1) -> Just South
-  V2 1 0 -> Just East
-  V2 (-1) 0 -> Just West
-  V2 0 0 -> Just Down
-  _ -> Nothing
+data DirInfo = DirInfo
+  { dirSyntax :: Text
+  , -- absolute direction if it exists
+    dirAbs :: Maybe (V2 Int64)
+  , -- the turning for the direction
+    dirApplyTurn :: V2 Int64 -> V2 Int64
+  }
 
--- | Convert a 'Direction' into a corresponding vector.  Note that
---   this only does something reasonable for 'North', 'South', 'East',
---   and 'West'---other 'Direction's return the zero vector.
-fromDirection :: Direction -> V2 Int64
-fromDirection d = case d of
-  North -> north
-  South -> south
-  East -> east
-  West -> west
-  _ -> V2 0 0
+allDirs :: [Direction]
+allDirs = [minBound .. maxBound]
+
+-- | TODO Information about all directions
+dirInfo :: Direction -> DirInfo
+dirInfo d = case d of
+  DLeft -> DirInfo directionSyntax Nothing (\(V2 x y) -> V2 (- y) x)
+  DRight -> DirInfo directionSyntax Nothing (\(V2 x y) -> V2 y (- x))
+  DBack -> DirInfo directionSyntax Nothing (\(V2 x y) -> V2 (- y) (- x))
+  DForward -> DirInfo directionSyntax Nothing id
+  DNorth -> DirInfo directionSyntax (Just north) (const north)
+  DSouth -> DirInfo directionSyntax (Just south) (const south)
+  DEast -> DirInfo directionSyntax (Just east) (const east)
+  DWest -> DirInfo directionSyntax (Just west) (const west)
+  DDown -> DirInfo directionSyntax (Just down) (const $ V2 0 0)
+ where
+  -- name is generate from Direction data constuctor
+  -- e.g. DLeft becomes "left"
+  directionSyntax = toLower . T.tail . from . show $ d
```

Comment:
```suggestion
dirInfo d = case d of
  DLeft -> relative (\(V2 x y) -> V2 (- y) x)
  DRight -> relative (\(V2 x y) -> V2 y (- x))
  DBack -> relative (\(V2 x y) -> V2 (- y) (- x))
  DForward -> relative id
  DNorth -> cardinal north
  DSouth -> cardinal south
  DEast -> cardinal east
  DWest -> cardinal west
  DDown -> cardinal down
 where
  -- name is generate from Direction data constuctor
  -- e.g. DLeft becomes "left"
  directionSyntax = toLower . T.tail . from . show $ d
  cardinal d = DirInfo directionSyntax (Just d) (const d)
  relative cd = DirInfo directionSyntax Nothing cd
```

Reply:
Cool, so this is how GitHub PR suggestion works :smile: Sorry about this perfectionism @twitu, I just wanted to try this out :wink: